core/src/main/java/io/grpc/ServerAddress.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2015, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import java.net.SocketAddress;
+
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A named server address with additional metadata for use in name resolution and load-balancing.
+ */
+@ExperimentalApi
+@Immutable
+public final class ServerAddress {
+  private final String name;
+  private final SocketAddress address;
+  private final Attributes attributes;
+
+  /**
+   * Constructor.
+   *
+   * @param name the name of this address
+   * @param address the address object
+   * @param attributes attributes associated with this address.
+   */
+  public ServerAddress(String name, SocketAddress address, Attributes attributes) {
+    this.name = name;
+    this.address = address;
+    this.attributes = attributes;
+  }
+
+  /**
+   * Returns the name of this address.
+   */
+  public String getName() {
```

Comment:
> Because ServerAddress also contains attributes. Either including or not the attributes for the equality of ServerAddress is problematic.

I fully agree. That is sort of why I'm scared of `name`, because it makes it very non-obvious whether or not attributes apply to equality.

If there was documentation about how names should include the "important" pieces (as defined by the creator of the address) and how you _must not_ compare SocketAddress names from different sources, then I could be more okay with that. Even with that though I could see it get annoying to produce a unique name for some attributes.